client/src/pages/Signup/components/SignupForm.tsx

```diff
@@ -0,0 +1,279 @@
+  id: string;
+  label: string;
+  value: string;
+  type?: string;
+  onChange: (event: React.ChangeEvent<HTMLInputElement>) => void;
+  onBlur: (event: React.FocusEvent<HTMLInputElement>) => void;
+  error?: boolean;
+  helperText?: string;
+}
+
+//=============== functions ===============//
+function delay(ms: number) {
+  return new Promise((resolve) => setTimeout(resolve, ms));
+}
+
+//=============== consts ===============//
+const validationSchema = yup.object().shape({
+  username: yup.string().required('Username cannot be empty'),
+  password: yup
+    .string()
+    .min(8, 'Password must be at least 8 characters')
+    .required('Password cannot be empty'),
+  confirmPassword: yup
+    .string()
+    .oneOf([yup.ref('password'), ''], 'Passwords must match')
+    .required('Confirm password cannot be empty'),
+  securityQuestion: yup.string().required('Security question cannot be empty'),
+  securityAnswer: yup
+    .string()
+    .required('Security answer answer cannot be empty'),
+});
+
+const SignupTextField: React.FC<SignupTextFieldProps> = ({
+  id,
+  label,
+  value,
+  type = 'text',
+  onChange,
+  onBlur,
+  error,
+  helperText,
+}) => {
+  return (
+    <TextField
+      required
+      id={id}
+      label={label}
+      variant="filled"
+      size="small"
+      type={type}
+      inputProps={{
+        style: { color: theme.palette.primary.contrastText },
+      }}
+      InputLabelProps={{
+        style: { color: theme.palette.primary.contrastText },
+      }}
+      style={{ width: 300 }}
+      value={value}
+      onChange={onChange}
+      onBlur={onBlur}
+      error={error}
+      helperText={helperText}
+    />
+  );
+};
+
+/**
+ * Signup Textfield
+ * @returns {JSX.Element} - Signup page
+ */
+const SignupForm = (): JSX.Element => {
+  const navigate = useNavigate();
+  const [usernameError, setUsernameError] = React.useState(' ');
+  const [generalError, setGeneralError] = React.useState(false);
+  const [openSnackbar, setOpenSnackbar] = React.useState(false);
+
+  const handleSignup = async (values: SignupFields) => {
+    try {
+      const { username, password, securityQuestion, securityAnswer } = values;
+      const response: AxiosResponse = await signup({
+        username,
+        password,
+        securityQuestion,
+        securityAnswer,
+      });
+      if (response.status === 201) {
+        setOpenSnackbar(true);
+        await delay(3000);
+        navigate('/login');
+      } else {
+        setGeneralError(true);
+      }
+    } catch (error) {
+      if (axios.isAxiosError(error)) {
+        const axiosError: AxiosError = error;
+        if (axiosError.response?.status === 400) {
+          setUsernameError('Username already exists');
+        } else {
+          setGeneralError(true);
+        }
+      } else {
+        setGeneralError(true);
+      }
+    }
+  };
+
+  const handleUsernameChange = (event: React.ChangeEvent<HTMLInputElement>) => {
+    setUsernameError(' ');
```

Comment:
is there a reason we set these errors to be just spaces instead of empty strings

Reply:
if the error is set to an empty string, the space for the error is not reserved and therefore when there is an error, all the components shift. quick fix is to just set it to a space.

Reply:
![image](https://github.com/BarbzCodez/Spendr/assets/64338275/818d59aa-7b7f-49b5-96cb-40193ab20fb3)
LGTM

maybe put comment or something to mention that since it's not obvious